### PR TITLE
DynamoDB: Implement multi-accounts/multi-region

### PR DESCRIPTION
**Pull request by viren-nadkarni**
### Summary
This PR implements multi-accounts and multi-region for DynamoDB.

DynamoDB depends on an external application -- DynamoDBLocal. Requests are forwarded to the DDBLocal instance, which has built-in support for AWS API. 

DDBLocal supports namespacing based on the value of `Credential` field in `Authorization` header. This is an undocumented feature. To achieve namespacing, we use a concatenation of account ID and region for this field before the request is forwarded to DDBLocal.

Furthermore, there are fixes to KMS and Kinesis integration. These are changed to use account ID and region from ARN instead of thread context. Unfortunately itself Kinesis does not fully support namespacing either given it also uses an external component. This will be addressed in a subsequent PR.

### Tests
New integration test which attempts to create table with same name in different region and different account.

### Related
Closes: #7126

### CODE REVIEW

File: localstack/utils/aws/aws_stack.py [hunk 1]
```diff
@@ -469,17 +469,6 @@ def inject_region_into_env(env, region):
...
-    client = client or connect_to_service("dynamodb")
-    paginator = client.get_paginator("list_tables")
-    pages = paginator.paginate(PaginationConfig={"PageSize": 100})
-    for page in pages:
-        table_names = page["TableNames"]
-        if to_str(table_name) in table_names:
-            return True
-    return False
-
-
```
Comment (viren-nadkarni) [inline localstack/utils/aws/aws_stack.py:482]: This function is simply moved together with other DynamoDB related functions

File: localstack/services/dynamodb/provider.py [hunk 2]
```diff
@@ -1111,11 +1171,10 @@ def fix_consumed_capacity(self, request: Dict, result: Dict):
...
 
-    def fix_table_arn(self, table_arn: str) -> str:
-        return re.sub(
-            "arn:aws:dynamodb:ddblocal:",
-            f"arn:aws:dynamodb:{aws_stack.get_region()}:",
-            table_arn,
+    def fix_table_arn(self, arn: str) -> str:
+        """Set the correct account ID and region in ARNs returned by DynamoDB local."""
+        return arn.replace(":ddblocal:", f":{aws_stack.get_region()}:").replace(
+            ":000000000000:", f":{get_aws_account_id()}:"
```
Comment (whummer) [inline localstack/services/dynamodb/provider.py:1177]: nit: same as above,  could use `DEFAULT_AWS_ACCOUNT_ID` here

**Review by whummer — APPROVED** [review]
Nice set of changes @viren-nadkarni ! 🚀 

Gave it a quick try locally - works nicely, also with persistence enabled. Restoring DDB tables does not seem to work with cloud pods yet - but that seems unrelated to this PR (also seems to be broken on `master`, afaict).

File: localstack/services/dynamodb/provider.py [hunk 3]
```diff
@@ -346,8 +353,8 @@ def _modify_ddblocal_arns(self, chain, context: RequestContext, response: Respon
         if response_content := response.get_data(as_text=True):
             # fix the table and latest stream ARNs (DynamoDBLocal hardcodes "ddblocal" as the region)
             content_replaced = re.sub(
-                r'("TableArn"|"LatestStreamArn"|"StreamArn")\s*:\s*"arn:([a-z-]+):dynamodb:ddblocal:([^"]+)"',
-                rf'\1: "arn:\2:dynamodb:{aws_stack.get_region()}:\3"',
+                r'("TableArn"|"LatestStreamArn"|"StreamArn")\s*:\s*"arn:([a-z-]+):dynamodb:ddblocal:000000000000:([^"]+)"',
```
[the remark this answers is not in the document]
Comment (viren-nadkarni) [reply localstack/services/dynamodb/provider.py:356]: I don't think it is correct to use `DEFAULT_AWS_ACCOUNT_ID` here. It represents the default & fallback account ID LocalStack uses. Whereas the `000000000000` is actually hardcoded in DDBLocal.

In the future we could change the value of `DEFAULT_AWS_ACCOUNT_ID` but DDBLocal output will continue to use `000000000000`.

**Comment by viren-nadkarni** [comment]
**ci/circleci: docker-test-arm64** All tests pass
```
============================= slowest 10 durations =============================
100.28s setup    tests/integration/test_serverless.py::TestServerless::test_apigateway_deployed
92.06s call     tests/integration/test_cloudwatch.py::TestCloudwatch::test_put_metric_alarm
46.74s call     tests/integration/test_firehose.py::TestFirehoseIntegration::test_kinesis_firehose_elasticsearch_s3_backup[domain]
39.60s setup    tests/integration/test_opensearch.py::TestOpensearchProvider::test_list_versions
38.27s call     tests/integration/test_firehose.py::TestFirehoseIntegration::test_delivery_stream_with_kinesis_as_source
35.41s call     tests/integration/test_firehose.py::TestFirehoseIntegration::test_kinesis_firehose_opensearch_s3_backup[path]
29.39s call     tests/integration/test_firehose.py::TestFirehoseIntegration::test_kinesis_firehose_opensearch_s3_backup[domain]
25.30s call     tests/integration/awslambda/test_lambda_whitebox.py::TestFunctionStates::test_invoke_failure_when_state_pending
24.89s call     tests/integration/test_opensearch.py::TestMultiClusterManager::test_multi_cluster
24.67s call     tests/integration/test_firehose.py::TestFirehoseIntegration::test_kinesis_firehose_elasticsearch_s3_backup[path]
= 422 passed, 88 skipped, 32 xfailed, 3 xpassed, 1570 warnings in 1538.23s (0:25:38) =
```

**Comment by giograno** [comment]
I just checked the restore of a pod after these changes. It works fine with the caveat that the `.db` assets now have both account and region names as suffixes, thus eventual pods need to be re-created to work after these changes!

**Pull request merged**